text/0000-nested-method-calls.md

```diff
@@ -0,0 +1,824 @@
+
+(Naturally, these complex arguments that are blocks look really
+artificial, but keep in mind that most of the time when this occurs in
+practice, the argument is a method or fn call, and that could in
+principle have arbitrary side-effects.)
+
+### Introducing reservations
+
+This RFC proposes extending MIR with the concept of a **two-phase
+borrow**. These borrows are a variant of mutable borrows where the
+value starts out as **reserved** and only becomes mutably borrowed
+when the resulting reference is first used (which is called
+**activating** the borrow). During the reservation phase before a
+mutable borrow is activated, it acts exactly like a shared borrow --
+hence the borrowed value can still be read.
+
+As discussed earlier, this RFC itself only introduces these two-phase
+borrows in a limited way. Specifically, we extend the MIR with a new
+kind of borrow (written `mut2`, for two-phase), and we generate those
+new kinds of borrows when lowering method calls.
+
+To understand how two-phased borrows help, let's revisit our two
+examples. We'll start with the motivating example,
+`vec.push(vec.len())`. When this expression is desugared, the
+resulting reference is stored into a temporary, `tmp0`.  Therefore,
+until `tmp0` is referenced again, `vec` is only considered
+**reserved**:
+
+```rust
+/* 0 */ tmp0 = &mut2 vec;       // reservation of `vec` starts here
+/* 1 */ tmp1 = &vec;
+/* 2 */ tmp2 = Vec::len(tmp1);
+/* 3 */ Vec::push(tmp0, tmp2); // first use of `tmp0`, upgrade is here
+```
+
+The first use of `tmp0` is on line 3, and hence the mutable borrow
+begins then, and lasts until the end of the borrow region. Crucially,
+lines 1 and 2 (which did a shared borrow of `vec`) took place during
+the reservation period, and hence no error results. This is because a
+reservation is equivalent to a shared borrow, and multiple shared
+borrows are allowed.
+
+Next, let's consider the sneaky example, where the argument attempts
+to mutate the vector that is being used in the receiver:
+
+```rust
+let mut v: Vec<String> = vec![format!("Hello, ")];
+let s: String = format!("foo");
+v[0].push_str({ v.push(s); "World!" });
+//              ^^^^^^^^^ sneaky attempt to mutate `v`
+```
+
+In this case, if we examine the resulting MIR, we can see that the
+borrow of `v` is almost **immediately** used, as part of the
+`IndexMut` operation:
+
+```rust
+// First evaluate `v[0]` to get a `&mut String`:
+tmp0 = &mut2 v;
+tmp1 = IndexMut::index_mut(tmp0, 0); // tmp0 used here!
+tmp2 = tmp1;
+
+// Next, evaluate `{ v.push(s); "World!" }` block:
+tmp3 = &mut2 v; // <-- Error! mutable borrow of `v` is active.
+... // see above
+```
+
+This implies that the mutable borrow will be active later on, when `v`
+is borrowed again during the arguments, and hence an error is still
+reported.
+
+Note that this same treatment will also rule out some "harmless"
+examples, such as this one:
+
+```rust
+v[0].push_str(&format!("{}", v.len()));
+```
```

Comment:
Indeed. I meant that it's "obviously" harmless in the sense that we know what those routines do (I can add a note to this effect).